### PR TITLE
EVG-7239 check instance type when user specified

### PR DESCRIPTION
**Pull request by ablack12**
Using EC2's new "describe instance type offerings," check if a user given instance type is both "allowed" via admin settings (this check was already true, I just moved it) and that the instance type is available for the current region.

### CODE REVIEW

File: cloud/ec2.go [hunk 1]
```diff
@@ -623,6 +623,25 @@ func (m *ec2Manager) setInstanceType(ctx context.Context, h *host.Host, instance
...
+		return errors.Wrapf(err, "error describe instance types offered for region '%s", m.region)
+	}
+	validTypes := []string{}
+	for _, availableType := range output.InstanceTypeOfferings {
+		if availableType.InstanceType != nil && (*availableType.InstanceType) == instanceType {
+			return nil
+		}
+		validTypes = append(validTypes, *availableType.InstanceType)
+	}
+	return errors.Errorf("available types for region '%s' are: %v", m.region, validTypes)
```
Comment (ablack12) [inline cloud/ec2.go]: potentially this error is TOO helpful: the result for us-east-1 for example is 
`problem contacting evergreen service: problem spawning host: 400 (Bad Request): available types for region 'us-east-1' are: [x1e.xlarge z1d.3xlarge r5.xlarge t2.micro r5d.2xlarge c5d.18xlarge m5.large m5ad.12xlarge c4.large g3.16xlarge c5d.metal z1d.xlarge i3en.metal c5.large m5d.2xlarge p3.8xlarge i3en.3xlarge m1.medium m5n.8xlarge z1d.6xlarge x1e.16xlarge c5d.9xlarge r5ad.12xlarge r5dn.24xlarge c5.24xlarge t3a.2xlarge i3en.large m5.metal m5.xlarge m5dn.24xlarge m5ad.24xlarge m5a.large c3.4xlarge i2.xlarge r3.large t3a.xlarge p3dn.24xlarge m5d.16xlarge t3.small t3.2xlarge c3.xlarge d2.8xlarge m5n.2xlarge r5dn.12xlarge t2.xlarge m5dn.4xlarge i3.8xlarge i3en.2xlarge g4dn.4xlarge r5.16xlarge c5d.large r5ad.large m5a.16xlarge r5ad.24xlarge r5.2xlarge g4dn.xlarge m5ad.4xlarge r5d.12xlarge m5a.24xlarge i3en.24xlarge r5.large m5d.4xlarge r5.8xlarge m5ad.2xlarge p2.xlarge m5n.16xlarge i2.4xlarge r4.8xlarge m5d.large g4dn.12xlarge f1.2xlarge m2.2xlarge r5dn.4xlarge inf1.6xlarge r5n.2xlarge m5a.4xlarge m3.medium m4.xlarge m4.large c5.4xlarge m1.xlarge r5.metal c5d.4xlarge a1.4xlarge m5.8xlarge c4.xlarge m5dn.2xlarge a1.xlarge m5dn.12xlarge r4.16xlarge m3.xlarge r3.xlarge m5n.large p2.8xlarge p3.2xlarge i3.large x1e.4xlarge f1.4xlarge r5n.12xlarge c5.xlarge a1.2xlarge cc2.8xlarge i3.4xlarge r5.4xlarge t3.xlarge t3a.medium r5.24xlarge i3.2xlarge r4.2xlarge a1.metal i3.metal m5d.8xlarge t3a.micro i3en.xlarge h1.16xlarge g4dn.16xlarge r3.2xlarge t3.micro d2.2xlarge g2.2xlarge i3en.6xlarge r4.xlarge m4.2xlarge i3en.12xlarge p3.16xlarge c1.medium r5ad.xlarge r5a.8xlarge g2.8xlarge t3.nano r5n.large m5.12xlarge i3.16xlarge x1e.2xlarge d2.4xlarge i2.8xlarge r5a.large m5.2xlarge m5a.12xlarge r5a.12xlarge m4.4xlarge m1.large m5a.xlarge c5d.12xlarge m2.4xlarge c5n.9xlarge c5n.18xlarge r5n.8xlarge g3.8xlarge c5n.large c5d.xlarge m5n.24xlarge r5.12xlarge m5d.24xlarge h1.8xlarge r5d.xlarge r5n.24xlarge z1d.large t1.micro i3.xlarge m5n.12xlarge r5a.16xlarge m5dn.16xlarge h1.2xlarge x1.16xlarge m2.xlarge r5d.4xlarge r5dn.16xlarge z1d.12xlarge r5d.16xlarge m5ad.xlarge t3a.small c5n.4xlarge m5d.metal c5.12xlarge r5a.xlarge h1.4xlarge c5d.24xlarge g4dn.8xlarge c5.2xlarge m5a.8xlarge m5d.12xlarge r5d.metal m5.4xlarge m3.large c3.8xlarge g3.4xlarge r5dn.xlarge m5n.xlarge inf1.2xlarge r5d.8xlarge t2.nano c3.large x1e.8xlarge t3.large inf1.xlarge c5.metal r5n.4xlarge m5.24xlarge c5n.xlarge r5n.xlarge r5dn.2xlarge p2.16xlarge r5n.16xlarge m3.2xlarge c5n.2xlarge f1.16xlarge r3.8xlarge m5.16xlarge m5dn.8xlarge m4.16xlarge r5d.large m5d.xlarge r5ad.4xlarge m5ad.large c5n.metal m5n.4xlarge t2.2xlarge c5d.2xlarge c5.18xlarge m5dn.large i2.2xlarge t3.medium m5a.2xlarge z1d.metal r5a.2xlarge m5dn.xlarge a1.medium c5.9xlarge c4.4xlarge m1.small c4.8xlarge t2.small m4.10xlarge r5a.4xlarge z1d.2xlarge r4.4xlarge g4dn.2xlarge t2.large g3s.xlarge r5d.24xlarge r5dn.large t3a.large x1.32xlarge c3.2xlarge x1e.32xlarge t2.medium inf1.24xlarge r3.4xlarge r5dn.8xlarge c4.2xlarge r5a.24xlarge a1.large d2.xlarge c1.xlarge t3a.nano r5ad.2xlarge r4.large]
`

File: cloud/ec2_fleet.go [hunk 2]
```diff
@@ -199,6 +199,25 @@ func (m *ec2FleetManager) GetInstanceStatus(ctx context.Context, h *host.Host) (
 	return status, nil
 }
 
+func (m *ec2FleetManager) CheckInstanceType(ctx context.Context, instanceType string) error {
```
Comment (ybrill) [inline cloud/ec2_fleet.go]: Could this be shared with ec2Manager (maybe in ec2_util.go)?

File: cloud/spawn.go [hunk 3]
```diff
@@ -172,6 +178,35 @@ func CreateSpawnHost(so SpawnOptions) (*host.Host, error) {
 	return intentHost, nil
 }
 
+func CheckInstanceTypeValid(ctx context.Context, d distro.Distro, requestedType string, allowedTypes []string) error {
+	if !IsEc2Provider(d.Provider) {
+		return errors.Errorf("cannot specify instance type for provider '%s'", d.Provider)
+	}
+	foundType := false
+	for _, allowedType := range allowedTypes {
```
Comment (ybrill) [inline cloud/spawn.go]: an alternative could be to use `util.StringSliceContains`

File: cloud/ec2.go (hunk 1 above)
Comment (ybrill) [reply cloud/ec2.go]: :wow:

File: cloud/spawn.go (hunk 3 above)
Comment (ablack12) [reply cloud/spawn.go]: Good call!

File: cloud/ec2_fleet.go (hunk 2 above)
Comment (ablack12) [reply cloud/ec2_fleet.go]: I think sharing this isn't worth it: the only part that could really be shared without type conversions is lines 211 to 218. Most functions it looks like are duplicated between the ec2FleetManager and ec2Manager for this same reason so I'm personally okay with keeping this as is.

Comment (ybrill) [reply cloud/ec2_fleet.go]: sgtm

File: cloud/ec2.go (hunk 1 above)
Comment (bsamek) [reply cloud/ec2.go]: Yeah, I worry a bit about log noise and large messages for users. Maybe we just say that "type <type> is invalid for region <region>", and leave it at that.

File: cloud/spawn.go [hunk 4]
```diff
@@ -172,6 +178,35 @@ func CreateSpawnHost(so SpawnOptions) (*host.Host, error) {
...
+		if requestedType == allowedType {
+			foundType = true
+			break
+		}
+	}
+	if !foundType {
+		return errors.New("This instance type has not been allowed by admins")
```
Comment (bsamek) [inline cloud/spawn.go]: Is this us admins or AWS admins?

Comment (ablack12) [reply cloud/spawn.go]: Us admins, I can specify that?

Comment (bsamek) [reply cloud/spawn.go]: I think it implies us. Wasn't sure if you meant AWS instead. LGTM as is.

**Pull request merged**